lib/ex_aws/config/auth_cache.ex

```diff
@@ -72,14 +70,35 @@ defmodule ExAws.Config.AuthCache do
           auth
 
         adapter ->
-          adapter.adapt_auth_config(auth, profile, expiration)
+          attemp_credentials_refresh(adapter, auth, profile, expiration)
       end
 
+    Process.send_after(self(), {:refresh_awscli_config, profile, expiration}, expiration)
     :ets.insert(ets, {{:awscli, profile}, auth})
 
     auth
   end
 
+  defp attemp_credentials_refresh(adapter, auth, profile, expiration, retries \\ 6) do
+    case adapter.adapt_auth_config(auth, profile, expiration) do
```

Comment:
In timing this out, what's the timeout on each request on the adapter?

1. We have a 30 second budget
2. If we divide that into six retries, we get 5 seconds for each request + wait
3. Each request is going to take X milliseconds (what's the upper bound on this?)
4. Each wait is going to be some random value between 0 and 5 seconds

Most of the time, we should be fine, but if latency is high enough then we'll burst our 30 second budget before exhausting retries, and thus never raise. I think we should _always_ raise.

Is there a way to intercept the genserver 30 second timeout and raise there instead?

Reply:
Another solution is to mark the time when we start retries and have a budget of, say, 25 seconds from that moment. Once time is up _or_ we've attempted X retries, we raise and give up.

Reply:
@brentjanderson if you have 6 tries with an average of 2.5 wait, the chances of getting out of the budget are pretty low right? it needs to happen that the process gets the 5 second wait, 6 times which seems highly unlikely.

I'm not sure if I follow you on "always raise" can you explain it more?

Reply:
@brentjanderson also from https://hexdocs.pm/elixir/1.12/GenServer.html#module-timeouts

```
If the process has no messages waiting when the timeout is set and the number of given milliseconds pass without any message arriving, then handle_info/2 will be called with :timeout as the first argument. The timeout is cleared if any message is waiting or arrives before the given timeout.
```

Reply:
I think of AWS taking 5 seconds for a while as a shooting star event, but I hear ya.
I like budgeting on time, but one issue here is that we cannot guess what will happen on the next request, we can only preemptively stop early, and even so, Amazon could take 10 seconds instead of 5...

Reply:
Alright. Let's not let perfect be the enemy of good here. 👍

Reply:
@brentjanderson we could improve this using `Task` and adding a timeout to the call so we have control on the maximum. We could add a max, say 3 seconds or bust. and then so now we have 15 seconds maximum if we have 5 retries, and the rest we can use to randomize.

This starts to get more complex, and not sure if it's worth the investment :D (really interesting though)

Reply:
`Task` would do. I'm good with whichever way you want to go. It's good enough for now to improve things tremendously. If we get bit we can revisit that issue when it arises.

Reply:
Agreed, we can take this for a spin, and see what happens, we can always make it more complex later 🙌